# HG changeset patch
# User alta88@fixall.com
# Date 1737060792 -3600
# Parent  f770dd0f4762537f93560ebf61214d5c5fe53394
Feed Subscribe dialog enhancements
NOTE: This UI should be rewritten/redesigned and remove xul/nsITree.

Authored by Alta88 and contributed to the Betterbird Project.
See https://github.com/Betterbird/thunderbird-patches/blob/main/LICENSE for license details.
All uses require attribution to the Author.

1. Improve tolerance of opml import from 3rd parties.
2. Opml file creation enhancements; save creator, feedType, folder outline type.
3. Subscribe dialog ui/visual improvements and regression fixes.
4. Enable feeditem move (default drag) and copy (ctrl after dragstart), cursor indicators.
5. Improve keyboard support
* * *
Some small fixes to subscribe dialog.

diff --git a/mail/themes/linux/mail/newsblog/feed-subscriptions.css b/mail/themes/linux/mail/newsblog/feed-subscriptions.css
--- a/mail/themes/linux/mail/newsblog/feed-subscriptions.css
+++ b/mail/themes/linux/mail/newsblog/feed-subscriptions.css
@@ -4,4 +4,9 @@
 
 /* ::::: Feed Subscription styling :::::: */
 
 @import url("chrome://messenger/skin/shared/feedSubscribe.css");
+
+#selectFolder {
+  margin-block: 2px;
+  padding-block: 4px !important;
+}
diff --git a/mail/themes/osx/mail/newsblog/feed-subscriptions.css b/mail/themes/osx/mail/newsblog/feed-subscriptions.css
--- a/mail/themes/osx/mail/newsblog/feed-subscriptions.css
+++ b/mail/themes/osx/mail/newsblog/feed-subscriptions.css
@@ -8,4 +8,9 @@
 
 #subscriptionsDialog {
   padding: 4px;
 }
+
+#selectFolder {
+  margin-block: 2px;
+  padding-block: 4px !important;
+}
diff --git a/mail/themes/shared/mail/feedSubscribe.css b/mail/themes/shared/mail/feedSubscribe.css
--- a/mail/themes/shared/mail/feedSubscribe.css
+++ b/mail/themes/shared/mail/feedSubscribe.css
@@ -14,8 +14,13 @@
   fill: color-mix(in srgb, var(--default) 20%, transparent);
   stroke: var(--default);
 }
 
+#rssSubscriptionsList[treelines="true"] > treechildren::-moz-tree-twisty {
+ /* The image's width is 12 pixels; accomodate treelines */
+  width: 16px;
+}
+
 #folderNameCol {
   flex: 2 2;
 }
 
diff --git a/mail/themes/windows/mail/newsblog/feed-subscriptions.css b/mail/themes/windows/mail/newsblog/feed-subscriptions.css
--- a/mail/themes/windows/mail/newsblog/feed-subscriptions.css
+++ b/mail/themes/windows/mail/newsblog/feed-subscriptions.css
@@ -4,4 +4,9 @@
 
 /* ::::: Feed Subscription styling :::::: */
 
 @import url("chrome://messenger/skin/shared/feedSubscribe.css");
+
+#selectFolder {
+  margin-block: 0;
+  padding-block: 2px !important;
+}
diff --git a/mailnews/extensions/newsblog/Feed.sys.mjs b/mailnews/extensions/newsblog/Feed.sys.mjs
--- a/mailnews/extensions/newsblog/Feed.sys.mjs
+++ b/mailnews/extensions/newsblog/Feed.sys.mjs
@@ -371,8 +371,26 @@ Feed.prototype = {
       aNewTitle
     );
   },
 
+  get feedType() {
+    return lazy.FeedUtils.getSubscriptionAttr(
+      this.url,
+      this.server,
+      "feedType",
+      ""
+    );
+  },
+
+  set feedType(feedType) {
+    lazy.FeedUtils.setSubscriptionAttr(
+      this.url,
+      this.server,
+      "feedType",
+      feedType
+    );
+  },
+
   get lastModified() {
     return lazy.FeedUtils.getSubscriptionAttr(
       this.url,
       this.server,
@@ -666,10 +684,13 @@ Feed.prototype = {
     if (aFeed.parseItems) {
       // Do this only if we're in parse/store mode.
       aFeed.removeInvalidItems(false);
 
-      if (aCode == lazy.FeedUtils.kNewsBlogSuccess && aFeed.mLastModified) {
-        aFeed.lastModified = aFeed.mLastModified;
+      if (aCode == lazy.FeedUtils.kNewsBlogSuccess) {
+        if (aFeed.mLastModified) {
+          aFeed.lastModified = aFeed.mLastModified;
+        }
+        aFeed.feedType = aFeed.mFeedType;
       }
 
       // Flush any feed item changes to disk.
       const ds = lazy.FeedUtils.getItemsDS(aFeed.server);
diff --git a/mailnews/extensions/newsblog/FeedUtils.sys.mjs b/mailnews/extensions/newsblog/FeedUtils.sys.mjs
--- a/mailnews/extensions/newsblog/FeedUtils.sys.mjs
+++ b/mailnews/extensions/newsblog/FeedUtils.sys.mjs
@@ -535,9 +535,11 @@ export var FeedUtils = {
     ds.data = ds.data.filter(x => x.url != aFeed.url);
     ds.saveSoon();
 
     // Update folderpane.
-    Services.obs.notifyObservers(aFeed.folder, "folder-properties-changed");
+    if (aFeed.folder) {
+      Services.obs.notifyObservers(aFeed.folder, "folder-properties-changed");
+    }
   },
 
   /**
    * Change an existing feed's url.
@@ -844,9 +846,9 @@ export var FeedUtils = {
     Services.obs.notifyObservers(aFolder, "folder-properties-changed");
 
     const win = Services.wm.getMostRecentWindow("Mail:News-BlogSubscriptions");
     if (win) {
-      win.FeedSubscriptions.mView.tree.invalidate();
+      win.FeedSubscriptions?.mView?.tree?.invalidate();
     }
   },
 
   /**
@@ -1808,9 +1810,9 @@ export var FeedUtils = {
           // Nice touch: notify so the window ca select the folder that now
           // contains the newly subscribed feed.
           // This is particularly nice if we just finished subscribing
           // to a feed URL that the operating system gave us.
-          Services.obs.notifyObservers(feed.folder, "folder-subscribed");
+          Services.obs.notifyObservers(feed.folder, "folder-attention");
 
           // Check for an existing feed subscriptions window and update it.
           const subscriptionsWindow = Services.wm.getMostRecentWindow(
             "Mail:News-BlogSubscriptions"
@@ -1847,9 +1849,9 @@ export var FeedUtils = {
           }
 
           feed.options = options;
           FeedUtils.setStatus(feed.folder, feed.url, "lastUpdateTime", now);
-        } else if (aDisable) {
+        } else if (!this.mSubscribeMode && aDisable) {
           if (
             Services.prefs.getBoolPref("rss.disable_feeds_on_update_failure")
           ) {
             // Do not keep retrying feeds with error states. Set persisted state
diff --git a/mailnews/extensions/newsblog/feed-subscriptions.js b/mailnews/extensions/newsblog/feed-subscriptions.js
--- a/mailnews/extensions/newsblog/feed-subscriptions.js
+++ b/mailnews/extensions/newsblog/feed-subscriptions.js
@@ -346,9 +346,9 @@ var FeedSubscriptions = {
     },
 
     getImageSrc(aRow) {
       const item = this.getItemAtIndex(aRow);
-      if ((item.folder && item.folder.isServer) || item.open) {
+      if (!item || item.folder?.isServer || item.open) {
         return "";
       }
 
       if (
@@ -483,9 +483,9 @@ var FeedSubscriptions = {
                 newItem.folder == curItem.parentFolder ||
                 newItem.folder.isServer)
             )
           ) {
-            dragDataResults.canDrop = true;
+            dragDataResults.canDrop = newItem.container;
           }
         }
       } else {
         // Try to get a feed url.
@@ -1082,9 +1082,11 @@ var FeedSubscriptions = {
     }
 
     const hideRec = recommendedUnitsVal.value == "";
     recommendedUnits.hidden = hideRec;
+    recommendedUnits.disabled = updateValue.disabled;
     recommendedUnitsVal.hidden = hideRec;
+    recommendedUnitsVal.disabled = updateValue.disabled;
 
     // Autotag items.
     const autotagEnable = document.getElementById("autotagEnable");
     const autotagUsePrefix = document.getElementById("autotagUsePrefix");
@@ -1111,9 +1113,9 @@ var FeedSubscriptions = {
 
     selectFolder.setAttribute("hidden", !aIsFeed);
     selectFolder._folder = aFolder;
     selectFolderValue.toggleAttribute("hidden", aIsFeed);
-    selectFolderValue.setAttribute("showfilepath", false);
+    selectFolderValue.removeAttribute("showfilepath");
 
     if (aIsFeed) {
       selectFolderPopup._ensureInitialized();
       selectFolderPopup.selectFolder(aFolder);
@@ -1125,22 +1127,31 @@ var FeedSubscriptions = {
       selectFolderValue.setAttribute("filepath", aFolder.filePath.path);
     }
   },
 
+  onBlurSelectFolderValue(aEvent) {
+    const target = aEvent.target;
+    target.removeAttribute("showfilepath");
+    target.value = target.getAttribute("prettypath");
+  },
+
   onClickSelectFolderValue(aEvent) {
     const target = aEvent.target;
     if (
       ("button" in aEvent &&
         (aEvent.button != 0 ||
-          aEvent.target.localName != "div" ||
+          aEvent.target.localName != "input" ||
           target.selectionStart != target.selectionEnd)) ||
       (aEvent.keyCode && aEvent.keyCode != aEvent.DOM_VK_RETURN)
     ) {
       return;
     }
 
     // Toggle between showing prettyPath and absolute filePath.
-    if (target.getAttribute("showfilepath") == "true") {
+    const showFilePath = target.getAttribute("showfilepath");
+    if (!showFilePath && "button" in aEvent) {
+      target.setAttribute("showfilepath", false);
+    } else if (target.getAttribute("showfilepath") == "true") {
       target.setAttribute("showfilepath", false);
       target.value = target.getAttribute("prettypath");
     } else {
       target.setAttribute("showfilepath", true);
@@ -1334,9 +1345,9 @@ var FeedSubscriptions = {
         this.mView.selection.tree.invalidateRow(
           this.mView.selection.currentIndex
         );
       }
-      if (aNode.id == "updateValue") {
+      if (["biffUnits", "updateValue"].includes(aNode.id)) {
         FeedUtils.setStatus(
           item.parentFolder,
           item.url,
           "updateMinutes",
@@ -1387,13 +1398,31 @@ var FeedSubscriptions = {
     return val ? Math.round(val * 1000) / 1000 : "";
   },
 
   onKeyPress(aEvent) {
+    if (aEvent.target.id == "subscriptionsDialog") {
+      // Close button.
+      return;
+    }
+    if (aEvent.keyCode == aEvent.DOM_VK_RETURN) {
+      if (aEvent.target.localName != "button") {
+        aEvent.preventDefault();
+      }
+      if (aEvent.target.id == "locationValue") {
+        if (!document.getElementById("addFeed").hidden) {
+          this.addFeed();
+        } else if (!document.getElementById("updateFeed").hidden) {
+          this.updateFeed();
+        }
+      }
+      return;
+    }
     if (
       aEvent.keyCode == aEvent.DOM_VK_DELETE &&
       aEvent.target.id == "rssSubscriptionsList"
     ) {
       this.removeFeed(true);
+      return;
     }
 
     this.clearStatusInfo();
   },
@@ -1402,8 +1431,9 @@ var FeedSubscriptions = {
     const item = this.mView.currentItem;
     this.updateFeedData(item);
     this.setFocus();
     this.updateButtons(item);
+    this.clearStatusInfo();
   },
 
   updateButtons(aSelectedItem) {
     const item = aSelectedItem;
@@ -1446,15 +1476,19 @@ var FeedSubscriptions = {
 
     const locationValue = document.getElementById("locationValue");
     const updateEnabled = document.getElementById("updateEnabled");
 
+    const updateValue = document.getElementById("updateValue");
+    const biffUnits = document.getElementById("biffUnits");
+
     const quickMode = document.getElementById("quickMode");
     const autotagEnable = document.getElementById("autotagEnable");
     const autotagUsePrefix = document.getElementById("autotagUsePrefix");
     const autotagPrefix = document.getElementById("autotagPrefix");
 
     const addFeedButton = document.getElementById("addFeed");
     const updateFeedButton = document.getElementById("updateFeed");
+    const addCertExceptionButton = document.getElementById("addCertException");
 
     const isServer = item.folder && item.folder.isServer;
     const isFolder = item.folder && !item.folder.isServer;
 
@@ -1463,13 +1497,10 @@ var FeedSubscriptions = {
       quickMode.disabled =
       autotagEnable.disabled =
         false;
 
-    updateEnabled.parentNode
-      .querySelectorAll("input,radio,label")
-      .forEach(item => {
-        item.disabled = !updateEnabled.checked;
-      });
+    updateValue.disabled = !updateEnabled.checked;
+    biffUnits.disabled = !updateEnabled.checked;
 
     autotagUsePrefix.disabled = !autotagEnable.checked;
     autotagPrefix.disabled =
       autotagUsePrefix.disabled || !autotagUsePrefix.checked;
@@ -1488,14 +1519,10 @@ var FeedSubscriptions = {
       quickMode.disabled =
         disable && !FeedUtils.getFeedUrlsInFolder(item.folder);
       // All other options disabled unless intent is to add a feed.
       updateEnabled.disabled = disable;
-      updateEnabled.parentNode
-        .querySelectorAll("input,radio,label")
-        .forEach(item => {
-          item.disabled = disable;
-        });
-
+      updateValue.disabled = disable;
+      biffUnits.disabled = disable;
       autotagEnable.disabled = disable;
 
       addFeedButton.disabled =
         addFeedButton != focusedElement &&
@@ -1508,9 +1535,9 @@ var FeedSubscriptions = {
       if (
         locationValue.value != item.url &&
         locationValue != document.activeElement &&
         focusedElement != updateFeedButton &&
-        focusedElement.id != "addCertException"
+        focusedElement != addCertExceptionButton
       ) {
         locationValue.value = item.url;
       }
       this.setPrefs(locationValue);
@@ -1916,8 +1943,10 @@ var FeedSubscriptions = {
   },
 
   onDragOver(aEvent) {
     this.mView._currentDataTransfer = aEvent.dataTransfer;
+    // Must prevent default. Otherwise dropEffect gets cleared.
+    aEvent.preventDefault();
   },
 
   mFeedDownloadCallback: {
     mSubscribeMode: true,
@@ -1927,8 +1956,9 @@ var FeedSubscriptions = {
       window.focus();
       // Feed is null if our attempt to parse the feed failed.
       let message = "";
       const win = FeedSubscriptions;
+      let feedUrl = feed.url;
       if (
         aErrorCode == FeedUtils.kNewsBlogSuccess ||
         aErrorCode == FeedUtils.kNewsBlogNoNewItems
       ) {
@@ -1946,8 +1976,9 @@ var FeedSubscriptions = {
           message = FeedUtils.strings.GetStringFromName(
             "subscribe-feedVerified"
           );
           win.updateStatusItem("statusText", message);
+          document.getElementById("locationValue").focus();
           return;
         }
 
         // Update lastUpdateTime if successful.
@@ -2025,8 +2056,11 @@ var FeedSubscriptions = {
             );
           }
 
           win.selectFeed(feed, parentIndex);
+          if (win.mActionMode == win.kSubscribeMode) {
+            win.mTree.focus();
+          }
         }
       } else {
         // Non success.  Remove intermediate traces from the feeds database.
         // But only if we're not in verify mode.
@@ -2036,8 +2070,11 @@ var FeedSubscriptions = {
           feed.url &&
           feed.server
         ) {
           FeedUtils.deleteFeed(feed);
+          // Do not setStatus() on feeds/folders if the feedUrl is invalid; it
+          // will create false entries in the db for an invalid url.
+          feedUrl = null;
         }
 
         if (aErrorCode == FeedUtils.kNewsBlogInvalidFeed) {
           message = FeedUtils.strings.GetStringFromName(
@@ -2071,18 +2108,24 @@ var FeedSubscriptions = {
         if (
           win.mActionMode == win.kUpdateMode ||
           win.mActionMode == win.kVerifyUrlMode
         ) {
-          document.getElementById("locationValue").focus();
+          const locationValue = document.getElementById("locationValue");
+          locationValue.value = feed.url;
+          locationValue.focus();
         }
       }
 
       win.mActionMode = null;
       win.clearStatusInfo();
       const code = feed.url.startsWith("http") ? aErrorCode : null;
       win.updateStatusItem("statusText", message, code);
-      FeedUtils.setStatus(feed.folder, feed.url, "code", aErrorCode);
-      FeedUtils.setStatus(feed.folder, feed.folder.URI, "code", "");
+      if (feed.folder) {
+        if (feedUrl) {
+          FeedUtils.setStatus(feed.folder, feedUrl, "code", aErrorCode);
+        }
+        FeedUtils.setStatus(feed.folder, feed.folder.URI, "code", "");
+      }
     },
 
     // This gets called after the RSS parser finishes storing a feed item to
     // disk.  aCurrentFeedItems is an integer corresponding to how many feed
@@ -2550,9 +2593,9 @@ var FeedSubscriptions = {
 
     return new Promise(resolve => {
       fp.open(rv => {
         if (rv != Ci.nsIFilePicker.returnOK || !fp.file) {
-          resolve(null);
+          resolve([null, null]);
           return;
         }
 
         this.opmlLastOpenDir = fp.file.parent;
@@ -2576,9 +2619,9 @@ var FeedSubscriptions = {
 
     if (this.mRSSServer.rootFolder.hasSubFolders) {
       const opmlDoc = document.implementation.createDocument("", "opml", null);
       const opmlRoot = opmlDoc.documentElement;
-      opmlRoot.setAttribute("version", "1.0");
+      opmlRoot.setAttribute("version", "2.0");
       opmlRoot.setAttribute("xmlns:fz", "urn:forumzilla:");
 
       this.generatePPSpace(opmlRoot, SPACES2);
 
@@ -2595,8 +2638,14 @@ var FeedSubscriptions = {
       this.generatePPSpace(head, SPACES4);
       const dt = opmlDoc.createElement("dateCreated");
       dt.appendChild(opmlDoc.createTextNode(new Date().toUTCString()));
       head.appendChild(dt);
+      this.generatePPSpace(head, SPACES4);
+      const ownerNameText =
+        FeedSubscriptions.brandShortName + " " + AppConstants.MOZ_APP_VERSION;
+      const ownerName = opmlDoc.createElement("ownerName");
+      ownerName.appendChild(opmlDoc.createTextNode(ownerNameText));
+      head.appendChild(ownerName);
       this.generatePPSpace(head, SPACES2);
       opmlRoot.appendChild(head);
 
       this.generatePPSpace(opmlRoot, SPACES2);
@@ -2708,9 +2757,10 @@ var FeedSubscriptions = {
       );
 
       // Make a folder outline element.
       folderOutline = parent.ownerDocument.createElement("outline");
-      folderOutline.setAttribute("title", folder.prettyName);
+      folderOutline.setAttribute("text", folder.prettyName);
+      folderOutline.setAttribute("fz:folder", true);
       this.generatePPSpace(parent, indentString(indentLevel + 2));
 
       if (folder.hasSubFolders) {
         FeedUtils.log.log(
@@ -2736,12 +2786,14 @@ var FeedSubscriptions = {
   },
 
   exportOPMLOutline(aFeed, aDoc) {
     const outRv = aDoc.createElement("outline");
-    outRv.setAttribute("type", "rss");
-    outRv.setAttribute("title", aFeed.title);
+    outRv.setAttribute(
+      "type",
+      aFeed.feedType.startsWith("ATOM") ? "atom" : "rss"
+    );
     outRv.setAttribute("text", aFeed.title);
-    outRv.setAttribute("version", "RSS");
+    outRv.setAttribute("version", aFeed.feedType);
     outRv.setAttribute("fz:quickMode", aFeed.quickMode);
     outRv.setAttribute("fz:options", JSON.stringify(aFeed.options));
     outRv.setAttribute("xmlUrl", aFeed.url);
     outRv.setAttribute("htmlUrl", aFeed.link);
@@ -2901,14 +2953,19 @@ var FeedSubscriptions = {
         const outlineName =
           outline.getAttribute("text") ||
           outline.getAttribute("title") ||
           outline.getAttribute("xmlUrl");
-        let feedUrl, folder;
-
-        if (outline.getAttribute("type") == "rss") {
-          // A feed outline.
-          feedUrl =
-            outline.getAttribute("xmlUrl") || outline.getAttribute("url");
+        let folder;
+
+        const isRssType = ["rss", "atom"].includes(
+          outline.getAttribute("type")
+        );
+        const feedUrl =
+          outline.getAttribute("xmlUrl") || outline.getAttribute("url");
+
+        if (isRssType || feedUrl) {
+          // A feed outline if it contains the above feedUrl attributes,
+          // regardless of whether isRssType is true (follows the spec).
           if (!feedUrl) {
             FeedUtils.log.warn(
               `importOPMLOutlines: skipping, type=rss <outline> has no url, outlineName ${outlineName}`
             );
@@ -2926,9 +2983,9 @@ var FeedSubscriptions = {
           }
 
           if (
             aParentNode.tagName == "outline" &&
-            aParentNode.getAttribute("type") != "rss"
+            !aParentNode.hasAttribute("xmlUrl")
           ) {
             // Parent is a folder, already created.
             folder = feedFolder;
           } else {
diff --git a/mailnews/extensions/newsblog/feed-subscriptions.xhtml b/mailnews/extensions/newsblog/feed-subscriptions.xhtml
--- a/mailnews/extensions/newsblog/feed-subscriptions.xhtml
+++ b/mailnews/extensions/newsblog/feed-subscriptions.xhtml
@@ -58,8 +58,13 @@
     <script>
       window.addEventListener("load", event => {
         FeedSubscriptions.onLoad();
       });
+      window.addEventListener("dialogaccept", event => {
+        if (!FeedSubscriptions.onDialogAccept()) {
+          event.preventDefault();
+        }
+      });
       window.addEventListener("keypress", event => {
         FeedSubscriptions.onKeyPress(event);
       });
       window.addEventListener("mousedown", event => {
@@ -156,19 +161,18 @@
                   id="nameValue"
                   type="text"
                   class="input-inline"
                   aria-labelledby="nameLabel"
-                  onchange="FeedSubscriptions.setPrefs(this);"
+                  oninput="FeedSubscriptions.setPrefs(this);"
                 />
                 <hbox class="input-container">
                   <html:input
                     id="locationValue"
                     type="url"
                     class="uri-element input-inline"
                     aria-labelledby="locationLabel"
                     placeholder="&feedLocation2.placeholder;"
-                    onchange="FeedSubscriptions.setPrefs(this);"
-                    onfocus="FeedSubscriptions.onFocusChange();"
+                    oninput="FeedSubscriptions.clearStatusInfo(); FeedSubscriptions.setFocus(this);"
                     onblur="FeedSubscriptions.onFocusChange();"
                   />
                   <hbox align="center">
                     <label
@@ -204,8 +208,9 @@
                     readonly="readonly"
                     aria-labelledby="feedFolderLabel"
                     onkeypress="FeedSubscriptions.onClickSelectFolderValue(event);"
                     onclick="FeedSubscriptions.onClickSelectFolderValue(event);"
+                    onblur="FeedSubscriptions.onBlurSelectFolderValue(event);"
                   />
                 </hbox>
               </vbox>
             </hbox>
@@ -283,9 +288,9 @@
                 id="autotagPrefix"
                 type="text"
                 class="input-inline"
                 placeholder="&autoTagPrefix.placeholder;"
-                onchange="FeedSubscriptions.setPrefs(this);"
+                oninput="FeedSubscriptions.setPrefs(this);"
               />
             </hbox>
             <separator class="thin" />
           </vbox>
@@ -294,9 +299,9 @@
         <hbox id="statusContainerBox" align="center">
           <vbox flex="1">
             <description id="statusText" />
           </vbox>
-          <spacer flex="1" />
+          <spacer flex="2" />
           <label
             id="validationText"
             collapsed="true"
             class="text-link"
diff --git a/mailnews/extensions/newsblog/test/unit/test_feedparser.js b/mailnews/extensions/newsblog/test/unit/test_feedparser.js
--- a/mailnews/extensions/newsblog/test/unit/test_feedparser.js
+++ b/mailnews/extensions/newsblog/test/unit/test_feedparser.js
@@ -125,9 +125,9 @@ add_task(async function test_download() 
             );
             return;
           }
           // Feed has downloaded - make sure the right type was detected.
-          Assert.equal(feed.mFeedType, test.feedType, "feed type matching");
+          Assert.equal(feed.feedType, test.feedType, "feed type matching");
           Assert.equal(feed.title, test.title, "title matching");
           // Make sure we're got the expected number of messages in the folder.
           const cnt = [...folder.messages].length;
           Assert.equal(cnt, test.expectedItems, "itemcount matching");
